File: timeline.py
import git
import os
import re
import pandas as pd
import matplotlib.pyplot as plt

repo = git.Repo(os.path.join(os.getcwd(), "notes"))

exFiles = ["inbox", "todo", "setup"]
exSubdirs = ["calendars"]
exExtensions = []


def excludeFiles(blob):
    name, _ = os.path.splitext(blob.name)
    return name in exFiles


def excludeSubdirs(blob):
    directory = os.path.dirname(blob.path)
    return directory in exSubdirs


def excludeExtensions(blob):
    _, ext = os.path.splitext(blob.name)
    return ext in exExtensions


excluders = [excludeFiles, excludeSubdirs, excludeExtensions]


def processContents(s):
    l1 = len(s)
    # Remove backslashes (weird escapes)
    s = re.sub(r"\\", "", s)
    l2 = len(s)

    # Remove blocks
    blockRe = r"#\+begin_?([^ \n]+).*?#\+end_?\1"
    s = re.sub(blockRe, "", s, flags=(re.I | re.S))
    l3 = len(s)

    # Remove keywords
    keywordRe = r"#\+(.*):(.*)\n"
    s = re.sub(keywordRe, "", s, flags=(re.I))
    l4 = len(s)

    print("len: {}, {}, {}, {}".format(l1, l2, l3, l4))
    return s


data = pd.DataFrame()


def extractData(commit):
    blobs = [
        x
        for x in c.tree.traverse()
        if x.type == "blob" and not any(f(x) for f in excluders)
    ]
    words = [len(processContents(b.data_stream.read().decode()).split())
             for b in blobs]
    return {
        "date": pd.to_datetime(c.authored_date, unit="s"),
        "size": sum(words),
        "files": len(blobs),
        "commit": commit.hexsha,
    }


for c in repo.iter_commits("master"):
    data = data.append(extractData(c), ignore_index=True)

data = data.set_index("date")

data["size"].plot()
data["files"].plot(secondary_y=True)
print(data.head())
plt.show()

print(data.iloc[data["size"].argmax()])

File: Timeline.jl
module NotesTimeline

using LibGit2
using LibGit2: GitCommit, GitRepo, GitRevWalker, GitTree, GitTreeEntry, GitBlob, with, filename, peel, treewalk, entrytype, isbinary, content
using DataFrames
using Dates
using Statistics
using Lazy

const EXCLUDE_FILES = Set(["inbox", "todo", "setup"])
const EXCLUDE_DIRS = Set(["calendars"])
const EXCLUDE_EXTS = Set([])

function excludeFiles(::Any, name, files=EXCLUDE_FILES)
    (base, _) = splitext(name)
    return base in files
end

function excludeSubdirs(path, ::Any, dirs=EXCLUDE_DIRS)
    return path in dirs
end

function excludeExtensions(::Any, name, exts=EXCLUDE_EXTS)
    (_, ext) = splitext(name)
    return ext in exts
end

const EXCLUDERS = Set([excludeFiles, excludeSubdirs, excludeExtensions])
exclude(path, name, excluders=EXCLUDERS) = any(e -> e(path, name), excluders)

function getFileContents(c::GitCommit)
    filecontents = @NamedTuple{file::String, content::String}[]
    treewalk(peel(c)) do path, te
        if entrytype(te) == GitBlob && !isbinary(GitBlob(te)) && !exclude(path, filename(te))
            push!(filecontents, (file=joinpath(path, filename(te)),
                                 content=content(GitBlob(te))))
        end
        return 0
    end
    return filecontents
end

function processCommit(id, repo)
    c = GitCommit(repo, id)
    contents = getFileContents(c)
    d = DataFrame(contents)
    d.commit = sprint(print, LibGit2.GitHash(c))
    return d
end

function date(id, repo)
    c = GitCommit(repo, id)
    sig = LibGit2.author(c)
    time = unix2datetime(sig.time + sig.time_offset*60)
end

date(ids::AbstractVector{String}, repo) = [date(id, repo) for id in ids]

function processRepo(path)
    r = GitRepo(path)
    @info "Loading $path repo"
    files = with(walker -> LibGit2.map(processCommit, walker), GitRevWalker(r)) |> dfs -> vcat(dfs...)
    transform!(
        files,
        :content => (x -> x .|> split .|> length) => :words,
    )

    commits = combine(
        DataFrames.groupby(files, :commit),
        nrow => :files,
        :words => sum,
    )
    transform!(
        commits,
        :commit => (ids -> date(ids, r)) => :time,
    )

    sort!(commits, [:time])
    @info ("Loaded $path repo with $(size(commits, 1)) commits and $(size(files, 1))"
           * " total files ($(commits[end, :files]) files in the most recent commit)")
    return commits, files
end

end
